sound/soc/sof/intel/tgl.c

```diff
@@ -20,6 +20,89 @@ static const struct snd_sof_debugfs_map tgl_dsp_debugfs[] = {
 	{"dsp", HDA_DSP_BAR,  0, 0x10000, SOF_DEBUGFS_ACCESS_ALWAYS},
 };
 
+static int tgl_dsp_core_get(struct snd_sof_dev *sdev, int core)
+{
+	struct sof_ipc_pm_core_config pm_core_config = {
+		.hdr = {
+			.cmd = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CORE_ENABLE,
+			.size = sizeof(pm_core_config),
+		},
+		.enable_mask = sdev->enabled_cores_mask | BIT(core),
+	};
+	int ret;
+
+	if (core > sdev->num_cores - 1) {
+		dev_err(sdev->dev, "invalid core id: %d for num_cores: %d\n", core,
+			sdev->num_cores);
+		return -EINVAL;
+	}
+
+	/* no need to keep ref count for primary core. It will not be powered off until D3 anyway */
+	if (core == SOF_DSP_PRIMARY_CORE)
+		return 0;
+
+	/* nothing to do if current core ref_count is > 0 */
+	if (sdev->dsp_core_ref_count[core]++ > 0)
+		return 0;
+
+	/* Now notify DSP */
+	ret = sof_ipc_tx_message(sdev->ipc, pm_core_config.hdr.cmd,
+				 &pm_core_config, sizeof(pm_core_config),
+				 &pm_core_config, sizeof(pm_core_config));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: enabling secondary core '%d' failed with %d\n",
+			core, ret);
+		--sdev->dsp_core_ref_count[core];
+	} else {
+		sdev->enabled_cores_mask |= BIT(core);
+	}
+
+	return ret;
+}
+
+static int tgl_dsp_core_put(struct snd_sof_dev *sdev, int core)
+{
+	struct sof_ipc_pm_core_config pm_core_config = {
+		.hdr = {
+			.cmd = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CORE_ENABLE,
+			.size = sizeof(pm_core_config),
+		},
+		.enable_mask = sdev->enabled_cores_mask & ~BIT(core),
+	};
+	int ret;
+
+	if (core > sdev->num_cores - 1) {
+		dev_err(sdev->dev, "invalid core id: %d for num_cores: %d\n", core,
+			sdev->num_cores);
+		return -EINVAL;
+	}
+
+	/* no need to do anything for primary core. It should not be powered off until D3 anyway */
+	if (core == SOF_DSP_PRIMARY_CORE)
+		return 0;
+
+	/* nothing to do if ref_count > 0 */
+	if (--sdev->dsp_core_ref_count[core])
+		return 0;
```

Comment:
I am a bit nervous about increments and decrements that are not protected by a mutex.

Reply:
@plbossart I dont feel the need for a mutex here. Please correct me if I am wrong but:
1. We dont have any mutex protecting core_power_up/down() today during probe/suspend/remove.
2. the new core_get/put during widget setup/free are not protected. But the same argument for not protecting the widget use_count applies here too isnt it? the core card mutex already protects against this.

Reply:
same, can this be moved to the higher level so that we only call platform-specific implementations when needed?